snapshot/merge.go

```diff
@@ -0,0 +1,269 @@
+		} else {
+			mergeKeys = append(mergeKeys, parentKey)
+		}
+	}
+
+	// Map each key in mergeKeys to each key making up the snapshot chain.
+	// applyChains' outer slice has the same order as parents, which is expected to be lowest merge input to highest.
+	// applyChains' inner slice is ordered from highest layer to lowest.
+	var applyChains [][]string
+	var baseKey string
+	for i, parent := range mergeKeys {
+		if i == 0 && sn.overlayBased {
+			// Overlay-based snapshotters can skip the base of the merge and just use it as the parent of the merge snapshot.
+			// Other snapshotters will start empty (with baseKey set to "")
+			baseKey = parent
+			continue
+		}
+		var chain []string
+		for curkey := parent; curkey != ""; {
+			info, err := sn.Stat(ctx, curkey)
+			if err != nil {
+				return errors.Wrapf(err, "failed to stat chain key %q", curkey)
+			}
+			chain = append(chain, info.Name)
+			curkey = info.Parent
+		}
+		// add an empty key as the bottom layer so that the real bottom layer gets diffed with it and applied
+		chain = append(chain, "")
+		applyChains = append(applyChains, chain)
+	}
+
+	// Make the snapshot that will be merged into
+	prepareKey := identity.NewID()
+	if err := sn.Prepare(ctx, prepareKey, baseKey); err != nil {
+		return errors.Wrapf(err, "failed to prepare %q", key)
+	}
+	applyMounts, err := sn.Mounts(ctx, prepareKey)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get mounts of %q", key)
+	}
+
+	tempLeaseCtx, done, err := leaseutil.WithLease(ctx, sn.lm, leaseutil.MakeTemporary)
+	if err != nil {
+		return errors.Wrap(err, "failed to create temporary lease for view mounts during merge")
+	}
+	defer done(context.TODO())
+
+	// externalHardlinks keeps track of which inodes have been hard-linked between snapshots (which is
+	//	enabled when sn.useHardlinks is set to true)
+	externalHardlinks := make(map[uint64]struct{})
+
+	// Iterate over (lower, upper) pairs in each applyChain, calculating their diffs and applying each
+	// one to the mount.
+	// TODO: note for @tonistiigi, this is where we are currently flattening input snapshots for both
+	// native and overlay implementations. What is the advantage of splitting this into separate snapshots
+	// for the overlay case?
+	for _, chain := range applyChains {
+		for j := range chain[:len(chain)-1] {
+			lower := chain[len(chain)-1-j]
+			upper := chain[len(chain)-2-j]
+
+			var lowerMounts Mountable
+			if lower != "" {
+				viewID := identity.NewID()
+				var err error
+				lowerMounts, err = sn.View(tempLeaseCtx, viewID, lower)
+				if err != nil {
+					return errors.Wrapf(err, "failed to get mounts of lower %q", lower)
+				}
+			}
+
+			viewID := identity.NewID()
+			upperMounts, err := sn.View(tempLeaseCtx, viewID, upper)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get mounts of upper %q", upper)
+			}
+
+			err = diffApply(tempLeaseCtx, lowerMounts, upperMounts, applyMounts, sn.useHardlinks, externalHardlinks, sn.overlayBased)
```

Comment:
I don't really like this. Running the filesystem-based scanner is too unstable (especially in the non-overlay case). If we are flattening (still not quite sure about that) then the layer diffs are only required when the whole merge parent can contain whiteouts(eg. exec can't). We should store the diff metadata externally for these cases (eg. fileop-rm, diffop, maybe tar extract), and just replay the changes here. If no whiteouts are allowed and flattening then probably don't need it at all. When differ is called once the diff metadata should be saved and be immutable. This can be follow-up but needs a todo for fixes.

Reply:
(I'm still working on the updates based on your other comments, just getting back on this one right away)

>  then the layer diffs are only required when the whole merge parent can contain whiteouts(eg. exec can't). 

Maybe I'm misunderstanding, but I don't think this is true. For example:
```
a := base.Run("touch /a1").Root()
b := base.Run("touch /a1").Root().Run("rm /a1").Root()
m := Merge(a,b)
```

In this case, if you join together the blob layers making up the merge and then unpack them, the deletion of `/a1` will affect the file created in both `a` and `b` and you end up with just `base`. So I think layer diffs are required for execs too if we want our exported merge images to be consistent. Let me know if I was misunderstanding what you were saying.

That actually gets at the overall point of using the filesystem based differ too. Our rule is that the mount for a merge is the same as unpacking the joined input blobs. That's what enables the optimizations of re-using blob inputs when exporting a merge and what allows you to get consistent results when you unpack an image created via merge-op on a runtime that isn't aware of snapshot merges (i.e. containerd).

So given that the exporter uses the filesystem differ, it makes sense to me that the implementation of merge-op would also use it because we want the result of both to look the same. To be clear, I'm not against your suggestions, just that I think our initial goal should be to stabilize+optimize the differ rather than move merge-op away from using it. 

> Running the filesystem-based scanner is too unstable (especially in the non-overlay case)
> When differ is called once the diff metadata should be saved and be immutable

I'm all for having follow-ups that make the filesystem differ more stable (let me know if you know of any test cases that could trigger this behavior so I'm at least aware of what the instability is) and for optimizing the differ by saving the metadata and re-using it whenever possible (both in merges and during exports).

Just for context though, based on my testing so far, while the current implementation in this PR is not optimized at all, it's still pretty fast. Merging 5 layers each with several thousands files+dirs in the diff (created by installing packages with very large dependency trees) was maxing out at 600ms for the duration of the snapshotter `Merge` method on my laptop.

I don't say this to claim we'll never benefit from optimizations like re-using diff metadata, I'm sure we will, but I do want to approach it cautiously and make sure the extra complication will be in the right place to actually make meaningful performance improvements.

Reply:
>  but I don't think this is true. For example:

This is true for the single last layer that is not possible atm. When it is on top of a base then there is no whiteout. Eg. `rm` when there was no file in base does not create an empty whiteout (but fileop.rm should). Anyway, I didn't suggest that we should have different handling based on what created immutable. When we need diff, it is fine but needs to only happen one time. And when we don't need it then we shouldn't have it.

> let me know if you know of any test cases that could trigger this behavior so I'm at least aware of what the instability is

Early versions of docker relied on diff stability for pull-push cycle. There were countless issues with that. People are creative and run in very weird systems.

Reply:
> (but fileop.rm should)

It doesn't right now though, right? You're just suggesting that we change it to work like that?

Overall, that all SGTM. I think we could approach it as making a thin layer on top of the existing differ interface that allows us to also handle persisting+reusing diff metadata and other optimizations.